qualtran/bloqs/phase_estimation/kaiser_window_state.py

```diff
@@ -0,0 +1,162 @@
+
+
+@attrs.frozen
+class KaiserWindowState(QPEWindowStateBase, GateWithRegisters):  # type: ignore[misc]
+    r"""Bloq to prepare a Kaiser window state for high confidence Quantum Phase Estimation.
+
+    Kaiser window states are optimal to minimize the probability of error outside a given
+    confidence interval.
+    Given the bitsize $m$ and parameter $\alpha$, the bloq prepares an $m$-bit state with
+    coefficients
+
+    $$
+        \sum\limits_{x=-M}^{M}\frac{1}{2M} \frac{I_0\left(\pi\alpha\sqrt{1-(x/M)^2}\right)}{I_0\left(\pi\alpha\right)}\ket{x}
+    $$
+
+    where $M = 2^{m-1}$. See Ref[1] for more details.
+
+
+    Args:
+        bitsize: Number of bits in the control register of QPE.
+        alpha: Shape parameter, determines trade-off between main-lobe width and side lobe level.
+
+    References:
+        [Analyzing Prospects for Quantum Advantage in Topological Data
+        Analysis](https://arxiv.org/abs/2209.13581), Appendix D
+    """
+
+    bitsize: SymbolicInt
+    alpha: float
+
+    @cached_property
+    def kaiser_state_coeff(self) -> np.ndarray:
+        if is_symbolic(self.bitsize):
+            raise ValueError(f"Cannot compute coefficients for symbilic {self}")
+        state = scipy.signal.windows.kaiser(2**self.bitsize, self.alpha * np.pi, sym=False)
+        return state / np.linalg.norm(state)
+
+    @property
+    def m_bits(self) -> SymbolicInt:
+        return self.bitsize
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([self.m_register])
+
+    @classmethod
+    def from_precision_and_delta(
+        cls, precision: SymbolicInt, delta: SymbolicFloat
+    ) -> 'KaiserWindowState':
+        r"""Estimate $\varphi$ to $precision$ bits with $1-\delta$ success probability.
+
+        See Eq.D14 and Eq.D15 of Ref[1] for more details.
+        $$
+            \alpha = \frac{1}{\pi}\ln{(\frac{1}{\delta})} + \mathcal{O}(\ln{\ln{\frac{1}{\delta}}})
+        $$
+        and
+        $$
+            m = n + \log_2{(\ln{\frac{1}{\delta}}} + \mathcal{O}(\frac{1}{2^n}\ln\ln{\frac{1}{\delta}}))
+        $$
+
+        Args:
+            precision: Number of bits of precision
+            delta: Probability of success.
+        """
+        alpha = 1 / pi(delta) * ln(1 / delta) + ln(ln(1 / delta))
+        m_bits = precision + ceil(log2(ln(1 / delta + 1 / 2**precision * ln(ln(1 / delta)))))
+        return KaiserWindowState(bitsize=m_bits, alpha=alpha)
+
+    def my_tensors(
+        self, incoming: Dict[str, 'ConnectionT'], outgoing: Dict[str, 'ConnectionT']
+    ) -> List['qtn.Tensor']:
+        import quimb.tensor as qtn
+
+        return [
+            qtn.Tensor(
+                data=self.kaiser_state_coeff.reshape((2,) * self.bitsize),
+                inds=[(outgoing['qpe_reg'], i) for i in range(int(self.bitsize))],
+                tags=[str(self)],
+            )
+        ]
+
+    def _kraus_(self) -> Iterable[np.ndarray]:
```

Comment:
why is this added? do we actually test it anywhere?

Reply:
The notebook uses `cirq.Simulator()` to sample from the resulting circuit. To enable the sampling, we need this.

Reply:
Refactored to do simulations using Qualtran and not rely on cirq simulator. Now inheriting from `Bloq` instead of `GateWithRegisters`.